tools/socktap/tcp_link.cpp

```diff
@@ -0,0 +1,160 @@
+
+    std::array<boost::asio::const_buffer, layers_> const_buffers;
+    for (auto& layer : osi_layer_range<OsiLayer::Link, OsiLayer::Application>()) {
+        const auto index = distance(OsiLayer::Link, layer);
+        packet->layer(layer).convert(tx_buffers_[index]);
+        const_buffers[index] = boost::asio::buffer(tx_buffers_[index]);
+    }
+
+    std::list<TcpSocket>::iterator i = sockets_.begin();
+
+    while (i != sockets_.end()) {
+        if ((*i).status() == TcpSocket::CONNECTED) {
+            (*i).request(const_buffers);
+            i++;
+        } else if ((*i).status() == TcpSocket::ERROR) {
+            sockets_.erase(i++);
+            std::cerr << "Socket removed" << std::endl;
+        } else {
+            i++;
+        }
+    }
+}
+
+void TcpLink::connect(ip::tcp::endpoint ep)
+{
+    if (callback_) {
+        sockets_.emplace_back(*io_service_, &callback_);
+        auto& sock = sockets_.back();
+        sock.connect(ep);
+    } else {
+        waiting_endpoints_.push_back(ep);
+    }
+}
+
+void TcpLink::accept(ip::tcp::endpoint ep)
+{
+    if (acceptors_.count(ep) == 0) {
+        acceptors_.insert({ep, ip::tcp::acceptor(*io_service_, ep)});
+    }
+
+    sockets_.emplace_back(*io_service_, &callback_);
+    auto& sock = sockets_.back();
+    sock.status(TcpSocket::ACCEPTING);
+
+    boost::system::error_code ec;
+    std::cout << "Accept connetions at " << ep.address().to_string() << ":" << ep.port() << std::endl;
+
+    acceptors_.find(ep)->second.async_accept(
+        sock.socket(),
+        boost::bind(
+            &TcpLink::accept_handler,
+            this,
+            ec,
+            ep,
+            &sock
+        )
+    );
+}
+
+void TcpLink::accept_handler(boost::system::error_code& ec, ip::tcp::endpoint ep, TcpSocket* sock)
+{
+    sock->status(TcpSocket::CONNECTED);
+    sock->do_receive();
+    accept(ep);
+}
+
+
+/**
+ * TcpSocket
+ */
+
+TcpSocket::TcpSocket(boost::asio::io_service& io_service, IndicationCallback* cb) :
+    socket_(io_service),
+    callback_(cb),
+    rx_buffer_(2560, 0x00)
+{
+}
+
+void TcpSocket::connect(ip::tcp::endpoint ep)
+{
+    boost::system::error_code ec;
+    socket_.connect(ep, ec);
+    if (!ec) {
+        status_ = CONNECTED;
+        do_receive();
+    }
+
+}
+
+void TcpSocket::request(std::array<boost::asio::const_buffer, layers_> const_buffers)
+{
+    boost::system::error_code ec;
+    socket_.write_some(const_buffers, ec);
+
+    if (ec) {
+        status_ = ERROR;
+    }
+}
+
+void TcpSocket::do_receive()
+{
+    socket_.async_read_some(
+        boost::asio::buffer(rx_buffer_),
+        boost::bind(
+            &TcpSocket::receive_handler,
+            this,
+            boost::placeholders::_1,
+            boost::placeholders::_2
+        )
+    );
+}
+
+void TcpSocket::receive_handler(boost::system::error_code ec, std::size_t length) {
+    if (!ec) {
```

Comment:
TCP streams have no message boundaries by design. Thus, we need to ensure that we take full packets out of the stream. One possibility would be a size field (of fixed size) at the start of each packet. The receiver can read as many bytes as indicated by this field and knows that the next read must be the size field for the follow-up packet.

Example: [uint16: 500] 500 bytes packet [uint16: 400] 400 bytes packet [uint16 ...] ...